### DISCUSSION TITLE
List pattern enhancement: `any` pattern

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/csharplang#6434] [Language Ideas]
## Brief Description of This Idea

I think "any" patterns are useful sometimes:

```csharp
// Use `Any` extension/instance method returning `bool`, passing with a predicate of element type.
if (collection.Any(static e => e is { Prop: >= 20 and not 100, Prop2: var str and not [] }))
    // ...

// Or just use a plain `foreach` loop or `for` loop, with a flag variable of type `bool` to achieve.
bool flag = false;
foreach (var e in collection)
{
    if (e is { Prop: >= 20 and not 100, Prop2: var str and not [] })
    {
        flag = true;
        break;
    }
}
if (flag)
    // ...
```

This can be simplified using pattern-matching syntax like:

```csharp
// Use `any` keyword to describe a pattern is bound with an element stored in the collection.
if (collection is any { Prop: >= 20 and not 100, Prop2: var str and not [] })
    // ...

// Or adjust `any` keyword to the first place,
// representing "any element in a collection satisfies a condition".
// Here the syntax adds an extra `in` keyword just because
// the sentence "any element in the collection is ..."
// (here element uses a discard) is intuitive and grammar-acceptable.
if (any _ in collection is { Prop: >= 20 and not 100, Prop2: var str and not [] })
    // ...

// Or just allow double `..` range tokens, indicating at least one element in the collection
// matches the specified pattern.
// Of course we should only allow two `..` tokens to be placed at the first
// and the last position in a list pattern,
// which means we cannot use `..` twice, putting them with adjacent places:
//     [.., .., pattern]: invalid
// or they are not placed at both the first and last places:
//     [.., pattern1, .., pattern2]: invalid
// other cases:
//     [.., ..]: invalid because I think it is unnecessary - no values are checked
//     [.., pattern1, pattern2, ..]: ??? will be discussed later
if (collection is [.., { Prop: >= 20 and not 100, Prop2: var str and not [] }, ..])
    // ...
```

which is the content what I want to post.

### Plain `any` Pattern

How about `any`-keyword-only pattern? I mean:

```csharp
if (collection is any)
   // ...
```

If C# allows this, what does it mean? I think that `is any` means:

* `collection is not null && collection.Count != 0` (For reference types)
* `collection.Count != 0` (For value types)

### Describe Something That `any` Pattern Cannot Be Described

The above syntax rules cannot describe a case that whether the variable being iterated satisfies a condition that cannot be described as a pattern being allowed in C# at present.

Here we just suppose a new syntax uses `when` clauses:

```csharp
if (collection is
        any { Prop: >= 42, Prop2: var str and not [] }
        when !string.IsNullOrWhitespace(str))
    // ...
```

Details:

```csharp
if (collection is
//                            Temporary variable
//                                     ↓
        any { Prop: >= 42, Prop2: var str and not [] }
//                                Use that variable
//                                       ↓
        when !string.IsNullOrWhitespace(str))
    // ...
```

### `all` Pattern

If C# allows "any" pattern, we can use it to describe an "all" pattern.

```csharp
//
// Option 1
//
// Use `not` + `any` pattern to describe a negation.
// The complex pattern means the condition returns true value if:
//    1) Variable `collection` is not a valid collection (e.g. `null`).
//    2) No elements in that collection can match that pattern.
// Please note that `not any pattern` is different with `any not pattern`.
// `not any pattern`: used for achieving `all` pattern
// `any not pattern`: used for checking whether at least one element mismatches the specified pattern
if (collection is not any pattern)
    // ...

//
// Option 2
//
// Or negate a pattern by inserting a `not` keyword before `any`,
// describing no such element can match the specified pattern
// (i.e. "any element in that collection matches the pattern" is not correct).
if (not any _ in collection is pattern)
    // ...

//
// Option 3
//
// Or just use a basic syntax to negate an extended list pattern.
if (collection is not [.., pattern, ..])
    // ...
```

### Should `[.., pattern1, pattern2, ..]` Be Legal?

If C# uses `[.., pattern, ..]` to describe an `any` pattern, so... Should `[.., p1, p2, ..]` be legal also?

We suppose that C# will treat `[.., p1, p2, ..]` as a `for` loop using indices to visit each element:

```csharp
bool flag = false;

//                             Ensure the boundary
//               because `e2` uses `i + 1` to get the next element to `e1`.
//                                      ↓
for (int i = 0; i < collection.Length - 1; i++)
{
    var e1 = collection[i];
    var e2 = collection[i + 1];
    if (e1 is p1 && e2 is p2)
    {
        flag = true;
        break;
    }
}
if (flag)
    // ...
```

This implementation is just like a normal `any` pattern rule.

Same cases like `[.., p1, p2, p3, ..]`, `[.., p1, _, p2, ..]` and so on. However, I think such cases are too complex. I'm unfamiliar with implementation a syntax feature using Roslyn, so I don't know how hard such feature can be achieved.

### `any` Pattern Limitations

From the above description about `any` pattern, I mentioned three different ways to achieve same meaning of an `any` pattern:

1. `instance.Any(predicate)`
2. `for` loop with a `bool` flag
3. `foreach` loop with a `bool` flag

Deeping in the complete code snippet, we found that the variable `collection` must be iterate-able:

* Either the type of `collection` implements `IEnumerable` or `IEnumerable<T>`
* Or the type contains a method called `GetEnumerator`, returning an instance, whose type:
  * Contains a property named `Current`
  * Contains a parameterless method named `MoveNext`, returning `bool`

However, if we choose the option that uses `for` loop to implement, `any` pattern will require an extra member: indexer.

By the way, if we choose `foreach` loop to achieve, patterns like `[.., p1, p2, ..]` cannot be legal because we cannot describe this pattern using single loop or nested `foreach` loops. These rules can be discussed later. I don't think I can give an exact and rigorous rule to limit on this complex case.

## Conclusion

I don't know whether I have enumerated all possible cases that this idea has been covered. Moreover, my English is not good I think, so some contents may not be described very well. If you find any wrong usages on English grammar and wording, please forgive me.

Any considerations on this? Thanks.

**Comment by CyrusNajmabadi** [comment]
Seems odd to have 'any', but not also have 'all'.

**Reply by jnm2** [reply]
The thing with `is [.., var foo, ..]` is whether it is defined as matching the first matching element and skipping all other matching elements, or last, etc.

**Comment by HaloFour** [comment]
IMO this sounds quite complicated for a single pattern and would perhaps be something better accomplished via active patterns where you can describe how you want the logic to be applied vs. having it baked into the language.

**Reply by kyoyama-kazusa** [reply]
Thanks.

**Comment by 333fred** [comment]
I personally don't think we'll have dedicated any/all patterns. Instead, I think we'll have user-defined patterns, https://github.com/dotnet/csharplang/issues/4131, which will enable users to define such a pattern on their own.

**Reply by kyoyama-kazusa** [reply]
Hi~ Thanks for your reply :)

I think that any and all patterns can be normalized because they can be translated into a loop or a method invocation, which is "under control" and can be more helpful to compilers to analyze the code (such as flow analysis and so on).

However, user-defined patterns uses methods to achieve. It does work, but the inner code in the method cannot be analyzed and checked when we use a pattern matching, which is more troublesome.
